src/DotNetty.Handlers/Tls/TlsHandler.Writer.cs

```diff
@@ -243,46 +256,90 @@ private void FinishWrap(byte[] buffer, int offset, int count, IPromise promise)
         }
 
 #if NETCOREAPP || NETSTANDARD_2_0_GREATER
-        private Task FinishWrapNonAppDataAsync(in ReadOnlyMemory<byte> buffer, IPromise promise)
+        private Task FinishWrapAsync(in ReadOnlyMemory<byte> memory, IPromise promise)
         {
             var capturedContext = CapturedContext;
-            Task future;
-            if (MemoryMarshal.TryGetArray(buffer, out var seg))
+            IByteBuffer buf;
+            if (memory.IsEmpty)
             {
-                future = capturedContext.WriteAndFlushAsync(Unpooled.WrappedBuffer(seg.Array, seg.Offset, seg.Count), promise);
+                buf = Unpooled.Empty;
             }
             else
             {
-                future = capturedContext.WriteAndFlushAsync(Unpooled.WrappedBuffer(buffer.ToArray()), promise);
+                buf = capturedContext.Allocator.Buffer(memory.Length);
+                buf.WriteBytes(memory);
             }
-            this.ReadIfNeeded(capturedContext);
+            var future = capturedContext.WriteAndFlushAsync(buf, promise);
+            this.ReadIfNeeded(capturedContext, isWrite: true);
             return future;
         }
 #endif
 
-        private Task FinishWrapNonAppDataAsync(byte[] buffer, int offset, int count, IPromise promise)
+        private Task FinishWrapAsync(byte[] buffer, int offset, int count, IPromise promise)
         {
             var capturedContext = CapturedContext;
-            var future = capturedContext.WriteAndFlushAsync(Unpooled.WrappedBuffer(buffer, offset, count), promise);
-            this.ReadIfNeeded(capturedContext);
+            IByteBuffer buf;
+            if (0 >= count)
+            {
+                buf = Unpooled.Empty;
+            }
+            else
+            {
+                buf = capturedContext.Allocator.Buffer(count);
+                buf.WriteBytes(buffer, offset, count);
+            }
+            var future = capturedContext.WriteAndFlushAsync(buf, promise);
+            this.ReadIfNeeded(capturedContext, isWrite: true);
             return future;
         }
         
 #if NETCOREAPP || NETSTANDARD_2_0_GREATER
-        private static async ValueTask LinkOutcome(ValueTask valueTask, IPromise promise)
+        private async ValueTask WriteAsync(IByteBuffer buf, IPromise promise)
         {
+            var lastAsyncWriteTask = _lastAsyncWriteTask;
+            if (lastAsyncWriteTask != null && !lastAsyncWriteTask.IsCompletedSuccessfully)
```

Comment:
do we want to await `lastAsyncWriteTask ` if it is not completed? Or this is intentional to see if it was not completed **successfully** and by doing `await lastAsyncWriteTask;` we throw

```suggestion
            if (lastAsyncWriteTask != null && !lastAsyncWriteTask.IsCompleted)
```

Reply:
Existing code is correct. There 3 possible task states:
1. pending
2. succeeded
3. failed
We only need to await in case of (1) (to make sure we don't run things in parallel) and (3) (to get exception and fail the write operation). In case of (2) we can immediately move to submitting next write.